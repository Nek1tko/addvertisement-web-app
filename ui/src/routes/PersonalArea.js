import React, { useState } from "react";
import { Box, Button, Typography } from "@material-ui/core";
import AuthService from '../services/auth.service';
import { Redirect } from "react-router-dom";
import TextField from '@mui/material/TextField';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogTitle from '@mui/material/DialogTitle';
import { Collapse } from "@material-ui/core";
import { Alert } from "@mui/material";

const PersonalArea = props => {
    const [name, setName] = useState('Test name');
    const [surname, setSurname] = useState('Test surname');
    const [open, setOpen] = useState(false);
    const [errorOpen, setErrorOpen] = useState(false);
    const [errorMessage, setErrorMessage] = useState('');

    const [editedName, setEditedName] = useState(name);
    const [editedSurname, setEditedSurname] = useState(surname);


    if (!AuthService.getUser()) {
        return <Redirect to="/login" />
    }

    const { phoneNumber, userId, } = AuthService.getUser();

    const handleClickOpen = () => {
        setOpen(true);
    };

    const handleCancelClose = () => {
        setOpen(false);
        setErrorOpen(false);
        setEditedName(name);
        setEditedSurname(surname);
    };

    const handleSaveClose = () => {
        setOpen(false);
        setErrorOpen(false);
        setName(editedName);
        setSurname(editedSurname);
    };

    return (
        <Box
            display='flex'
            flexDirection='column'
            justifyContent='space-between'
            sx={{ mt: 10, width: '25%', m: 'auto' }}
        >
            <Typography variant="h4" align="center">
                {name}
            </Typography>

            <Typography variant="h4" align="center" style={{ marginTop: 30 }}>
                {surname}
            </Typography>

            <Typography variant="h5" align="center" style={{ color: "#666666", marginTop: 30 }}>
                {phoneNumber}
            </Typography>

            <Button
                variant="contained"
                color="primary"
                fullWidth
                mt={20}
                style={{ textTransform: 'none', marginTop: 30 }}
                onClick={handleClickOpen}
            >
                Редактировать
            </Button>

            <Dialog open={open} onClose={handleCancelClose}>
                <DialogTitle>Редактирование имени и фамилии</DialogTitle>

                <DialogContent>
                    <DialogContentText>
                        Введите имя и фамилию
                    </DialogContentText>

                    <Collapse in={errorOpen}>
                        <Alert
                            severity="error"
                            sx={{ mb: 2 }}
                        >
                            {errorMessage}
                        </Alert>
                    </Collapse>

                    <TextField
                        label="Имя"
                        variant="filled"
                        value={editedName}
                        onChange={e => {
                            const value = e.target.value;
                            if (value.length < 3 || value.length > 50) {
                                setErrorOpen(true);
                                setErrorMessage("Длина имени должна быть больше 4 и меньше 50");
                            }
                            else {
                                setEditedName(value);
                                setErrorOpen(false);
                            }
                        }}
                        fullWidth
                    />

                    <TextField
                        label="Фамилия"
                        variant="filled"
                        value={editedSurname}
                        onChange={e => {
                            const value = e.target.value;
                            if (value.length < 3 || value.length > 50) {
                                setErrorOpen(true);
                                setErrorMessage("Длина фамилии должна быть больше 4 и меньше 50");
                            }
                            else {
                                setEditedSurname(value);
                                setErrorOpen(false);
                            }
                        }}
                        fullWidth
                    />
                </DialogContent>

                <DialogActions>
                    <Button onClick={handleCancelClose}>Отмена</Button>
                    <Button onClick={handleSaveClose}>Сохранить</Button>
                </DialogActions>
            </Dialog>
        </Box>
    );
};

export default PersonalArea;
