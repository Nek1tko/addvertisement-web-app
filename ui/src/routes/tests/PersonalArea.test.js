/**
 * @group unit
 */

import React from "react";
import AuthService from "../../services/auth.service";
import { act } from "react-dom/test-utils";
import { mount, shallow } from "enzyme";
import { Router } from "react-router-dom";
import PersonalArea from "../PersonalArea";
import { createBrowserHistory } from "history";
import axios from "axios";

jest.mock('axios');
jest.mock("../../services/auth.service");

const mockUserAxios = {
    data: {
        name: "mockName",
        surname: "mockSurname",
    }
}

const mockUserAuthService = {
    userId: "mockId",
    phoneNumber: "mockPhoneNumber",
}

const eventName = {
    target: {
        value: "mockName"
    }
};

const eventName2 = {
    target: {
        value: "m"
    }
};


describe('PersonalAreaTests', () => {
    let wrapper, props;
    const setState = jest.fn();

    beforeEach(() => {
        props = {
            history: {
                ...createBrowserHistory(),
                push: jest.fn()
            }
        };
        AuthService.getUser.mockReturnValue(mockUserAuthService);
        axios.get.mockResolvedValue(mockUserAxios);
        axios.put.mockResolvedValue(mockUserAxios);
        const useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation((state) => [state, setState]);
        wrapper = mount(
            <Router history={props.history}>
                <PersonalArea />
            </Router>);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });


    test('PersonalAreaRenderTest', () => {
        act(() => {
            mount(<Router history={props.history}>
                <PersonalArea />
            </Router>);
        });
    });

    test('PersonalAreaSnapshotTest', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('PersonalAreaOpenButtonTest', async () => {
        await act(async () => {
            await wrapper.find('#openButton').first().props().onClick();
        });
        expect(setState).toHaveBeenCalledTimes(5);
    });

    test('PersonalAreaCloseButtonTest', async () => {
        await act(async () => wrapper = await shallow(<PersonalArea />));
        jest.clearAllMocks();

        const dived = wrapper.dive();
        await act(async () => {
            await dived.find('#closeButton').first().props().onClick();
        });

        expect(setState).toHaveBeenCalledTimes(4);
    });

    test('PersonalAreaSaveButtonTest', async () => {
        await act(async () => wrapper = await shallow(<PersonalArea />));
        jest.clearAllMocks();

        const dived = wrapper.dive();
        await act(async () => {
            await dived.find('#saveButton').first().props().onClick();
        });

        expect(setState).toHaveBeenCalledTimes(4);
    });

    test('PersonalAreaPositiveChangeNameTextFieldTest', async () => {
        await act(async () => wrapper = await shallow(<PersonalArea />));
        jest.clearAllMocks();

        const dived = wrapper.dive();
        await act(async () => {
            const nameTextField = await dived.find('#nameTextField');
            await nameTextField.props().onChange(eventName);
        });

        expect(setState).toHaveBeenCalledTimes(2);

    });

    test('PersonalAreaNegativeChangeNameTextFieldTest', async () => {
        await act(async () => wrapper = await shallow(<PersonalArea />));
        jest.clearAllMocks();

        const dived = wrapper.dive();
        await act(async () => {
            const nameTextField = await dived.find('#nameTextField');
            await nameTextField.props().onChange(eventName2);
        });

        expect(setState).toHaveBeenCalledTimes(2);
    });

    test('PersonalAreaPositiveChangeSurnameTextFieldTest', async () => {
        await act(async () => wrapper = await shallow(<PersonalArea />));
        jest.clearAllMocks();

        const dived = wrapper.dive();
        await act(async () => {
            const nameTextField = await dived.find('#surnameTextField');
            await nameTextField.props().onChange(eventName);
        });

        expect(setState).toHaveBeenCalledTimes(2);

    });

    test('PersonalAreaNegativeChangeSurnameTextFieldTest', async () => {
        await act(async () => wrapper = await shallow(<PersonalArea />));
        jest.clearAllMocks();

        const dived = wrapper.dive();
        await act(async () => {
            const nameTextField = await dived.find('#surnameTextField');
            await nameTextField.props().onChange(eventName2);
        });

        expect(setState).toHaveBeenCalledTimes(2);
    });

    test('PersonalAreaNotAuthorizedTest', () => {
        AuthService.getUser.mockReturnValue(null);
        wrapper = mount(
            <Router history={props.history}>
                <PersonalArea />
            </Router>);
    });
});
