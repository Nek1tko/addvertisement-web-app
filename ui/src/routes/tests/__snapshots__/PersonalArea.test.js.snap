// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalAreaTests PersonalAreaSnapshotTest 1`] = `ReactWrapper {}`;
